### ISSUE TITLE
Read dataset function: setBackground in Utils.py

### ISSUE THREAD

**Issue by Hugo101** [starling-lab/rfgb.py#10]
Take the Dataset ToyCancer as an example. There are 3 literals:

         - friends(+person,-person)
         - friends(-person,+person)
         - smokes(+person)

However, the function: **setBackground** in _Utils.py_ could miss some literals in the dataset. The following are the results of this function.

      - literalName: friends,   literalTypeSpecification: ['+person', '-person']
      - literalName: friends,   literalTypeSpecification: ['-person', '+person']
      - literalName: smokes,    literalTypeSpecification: ['+person']
      - final literals: {'friends': ['-person', '+person'], 'smokes': ['+person']}

The final literals have only two literals which misses **the first one:**
      

      - friends(+person,-person)

The reason why it happens is that it uses **dictionary** data structure. Some value is overridden because of the same key. 
We could apply **list** instead of **dict** to resolve it.

**Comment by batflyer** [comment]
I've narrowed this issue down to a few functions and classes: `utils.Data`, `utils.setBackground`, `utils.addVariableTypes`, `tree.expandOnBestTest`. So long as these are handled properly, the underlying `logic` implementation *shouldn't* need to change.

Starting from `utils.Data` and `utils.setBackground`:

```diff
class Data(object):
    ...
    def __init__(self, regression=False, advice=False,
                 softm=False, alpha=0.0, beta=0.0):
    ...
-   self.literals = {}
+   self.literals = []
    ...
    def setBackground(self, bk):
        ...
        for literalBk in bkWithoutTargets:
            literalName = literalBk.split('(')[0]
            literalTypeSpecification = literalBk[:-1].split('(')[1].split(',')
+           self.literals.append([literalName, literalTypeSpecification])
-           self.literals[literalName] = literalTypeSpecification
```

Changing data structure from dictionary to list will require updating how the literals are accessed in `tree.expandOnBestTest`:

```diff
class node(object):
    ...
    def expandOnBestTest(self, data=None):
        ...
        literals = data.getLiterals()
        tests = []
        for literal in literals:
+          literalName = literal[0]
+           literalTypeSpecification = literal[1]
-           literalName = literal
-           literalTypeSpecification = literals[literal]
```

`utils.addVariableTypes` will need to be updated in a similar manner, but in this case it makes sense to use a dictionary. This function updates Utils.data.variableType so that each argument in a literal may be mapped back to its type (e.g. `smokes(A)` -> `{'O': 'person'}`).

**Comment by Hugo101** [comment]
Thank you so much! Alex!

**Comment by batflyer** [comment]
Merged in at 53518e161dde9957c7ab10e319b578823f6b97ac, setting issue as resolved.
